packages/plugin-koa/src/koa.js

```diff
@@ -0,0 +1,64 @@
+const createReportFromErr = require('@bugsnag/core/lib/report-from-error')
+const extractRequestInfo = require('./request-info')
+
+const handledState = {
+  severity: 'error',
+  unhandled: true,
+  severityReason: {
+    type: 'unhandledErrorMiddleware',
+    attributes: { framework: 'Koa' }
+  }
+}
+
+module.exports = {
+  name: 'koa',
+  init: client => {
+    const requestHandler = async (ctx, next) => {
+      // Start a session whether sessions are used or not. We use this
+      // to store request-specific info, in case of any errors.
+      const sessionClient = client.startSession()
+
+      ctx.bugsnag = sessionClient
+
+      // extract request info and pass it to the relevant bugsnag properties
+      const requestInfo = extractRequestInfo(ctx)
+      sessionClient.metaData = { ...sessionClient.metaData, request: requestInfo }
+      sessionClient.request = {
+        clientIp: requestInfo.clientIp,
+        headers: requestInfo.headers,
+        httpMethod: requestInfo.httpMethod,
+        url: requestInfo.url,
+        referer: requestInfo.referer
+      }
```

Comment:
Do we need to do all this work when there will be potentially not be an error that we need to send this information for? Would it be better to gather all this in the catch below?

Reply:
This is a good point and I did consider it. I left it like this because the original bugsnag-node behaved in this way. The only downside I see is that in the case of an error some data might not be available? Otherwise it seems like the better thing to do.

Reply:
Ahhh I just took a deeper look at this.

The reason the data is attached to the `client` at this point is so that the user can call `req.bugsnag.notify()` and it has all of the data already captured. We could solve this using some higher level notifier feature like `beforeNotify` (as opposed to `beforeSend`) callbacks, to only attach the data when it is required. But really this is not much work in the grand scheme of things so I don't think it is a performance concern.